File: 46.permutations.js
/*
 * @lc app=leetcode id=46 lang=javascript
 *
 * [46] Permutations
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  const helper = (cur, idx) => {
    if (idx === nums.length) {
      res.push(cur.slice());
      return;
    }
    for (let i = 0; i < nums.length; ++i) {
      if (!visited[i]) {
        visited[i] = true;
        cur.push(nums[i]);
        helper(cur, idx + 1);
        cur.pop();
        visited[i] = false;
      }
    }
  };
  const res = [];
  const visited = new Array(nums.length).fill(false);
  helper([], 0);
  return res;
};
// @lc code=end

File: 46.permutations.py
#
# @lc app=leetcode id=46 lang=python3
#
# [46] Permutations
#

# Time:  O()
# Space: O()

from typing import List


# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def helper(cur: list, idx):
            if len(cur) == len(nums):
                res.append(cur[:])
                return
            for i in range(len(nums)):
                if visited[i]:
                    continue
                visited[i] = True
                cur.append(nums[i])
                helper(cur, idx + 1)
                cur.pop()
                visited[i] = False

        res, visited = [], [False] * len(nums)
        helper([], 0)
        return res


# @lc code=end
